powerquery-docs/query-folding-examples.md

```diff
@@ -0,0 +1,290 @@
+
+Transforms can roughly be grouped into two categories:
+
+- **Full scan operators**: These are operators that need to gather all the rows before the data can move on to the next operator in the chain. For example, in order to sort data, Power Query needs to gather all the data. Other examples of Full scan operators are *Table.Group*, *Table.NestedJoin*, *Table.Pivot*.
+
+-	**Streaming operators**: These are pass-through operators. For example, Table.SelectRows with a simple filter can usually filter the results as they pass through the operator, and won’t need to gather all rows before moving the data. *Table.SelectColumns*, T*able.ReorderColumns*, *Table.LastN* are additional examples of this sort of operators.
+
+
+>[!TIP]
+>While not every single transform requires the same resources to compute its resources, it is often more optimal to have a small number of nodes to show in the query plan. 
+
+## Considerations and suggestions
+
+* Follow the best practices when creating a new query as stated in [Best practices in Power Query](best-practices.md).
+* Use the step folding indicators to check which steps are preventing your query from folding and re-order them if necessary to increase folding.
+* Use the query plan to determine which transforms are happening at the Power Query engine for that particular step and consider re-arranging your query to check the updated query plan. For data sources that support folding, any nodes in the query plan other than Value.NativeQuery and data source access nodes represent transforms that didn’t fold.
+* Check the **View Native Query** option is always recommended to make sure that your query can be folded back to the data source. If your step disables this option, you know that you've created a step that stops query folding. Note that not all sources support this functionality, in those cases, you can rely on the step folding indicators and query plan.
+* Use the query diagnostics tools to your advantage and to better understand the requests being sent to your data source when query folding capabilities are available for the connector.
+* When combining data sourced from the use of multiple connectors, Power Query tries to push as much work as possible to both of the data sources while complying with the privacy levels defined for each data source. 
+* Read the article on [Privacy levels](dataprivacyfirewall.md) to protect your queries from running against a Data Privacy Firewall error.
+* You can also use other tools to check query folding from the perspective of the request being received by the data source. Based on our example, you can use the Microsoft SQL Server Profile to check the requests being sent by Power Query and received by the Microsoft SQL Server.
+* Whenever you add a new step to your query, Power Query will try to fold it back to the data source as a new data source query that will be evaluated by your data source.
```

Comment:
This doesn't really explain the potential issue that this can cause (doing something trivial on a small result triggering a long refresh time in the preview).

Reply:
I believe there might be an opportunity here to tackle a much broader topic on "how things get evaluated for the data preview" which is different than the evaluation that happens for a full query execution (like in SSIS for example).

Do you think that we should rephrase this sentence to be more explicit about the scenario that you mention? or should we try and tackle this and other scenarios related to how the data preview gets evaluated and cached in a separate article? I just want to make sure that when an article tackles a topic, that most of the follow-up questions to such topic are also addressed somewhere in the docs.

Reply:
While something broader would be fine, I think all that's necessary here is to clarify the potential pitfall in this sentence.

Reply:
That would work. Do you think that you could add your suggestions to the article? I've been hitting my head against the wall without having a clear idea to showcase that pitfall with just words.

Reply:
Ok. How about this?

```suggestion
* If you add a new step to a fully folded query and the new step also folds, Power Query may send a new request to the data source, rather than using a cached version of the previous result. In practice, this can result in seemingly simple operations on a small amount of data taking longer to refresh in the preview than expected, due to the fact that Power Query is re-querying the data source rather than working off a local copy of the data.
```

Reply:
LGTM!